### DISCUSSION TITLE
Generic value parameters

### DISCUSSION THREAD

**Discussion by OdinAnDieFreude** [dotnet/csharplang#5363] [Language Ideas]
The title says it all.

I think it would be a great addition to the type system, if a type could be parameterized not only by type parameters, but also by passed values.

but of couse, for all types to be known at compile time, the passed generic value parameters would have to be known at compile time.

    public interface IAdd<T>
    {
      T Add(T other);
    }
    public class X : IAdd<X>
    {
      public X Add(X other) { ... }
    }

    public class Matrix<T, int w, int h> : IAdd<Matrix<T, int w, int h>>
    {
      private T[,] Data = new T[w, h];
      public T this[int i, int j]
      {
        get => Data[i, j];
        set => Data[i, j] = value;
      } 
      public Matrix<T, w, h> Add(Matrix<T, w, h> other)
      {
        var result = new Matrix<T, w, h>();
        for (int j = 0; j < h; j++)
          for (int i = 0; i < w; i++)
            result[i, j] = Data[i, j] + other[i, j];
        return result;
      }
    }

    public void SomeMethod()
    {
      Matrix<X, 13, 34> a = ...;
      Matrix<X, 13, 34> b = ...;
      var c = a.Add(b);
      Matrix<X, 12, 34> d = ...;
      var e = a.Add(d); //compiler error
    }

And some ideas to expand on that concept:

params generic value parameters

    public class Matrix<T, params int[] dim> : IAdd<Matrix<T, dim>>
    {
      ...
    }

Calculating with generic value parameters, as long as the resulting value is known at compile time:

    class Polynomial<int degree>
    {
      private double[] Data = new double[degree + 1];
      public double this[int j]
      {
        get => Data[j];
        set => Data[j] = value;
      }
      public Polynomial<degree + d> Multiply<int d>(Polynomial<d> other)
      {
        var result = new Polynomial<degree + d>();
        for (int i = 0; i < degree; i++)
          for (int j = 0; j < d; j++)
            result[i + j] += Data[i] * other[j];
        return result;
      }
    }

**Comment by orthoxerox** [comment]
That's dependent typing, isn't it?

**Reply by OdinAnDieFreude** [reply]
Seems like it, I didn't know the term up until right now.

**Comment by Richiban** [comment]
Read up on Idris or F* (a research variant of F#) for languages that support dependent types, i.e. types that can depend on values.

IIRC Rust supports a tiny bit of this with respect to array lengths.

**Reply by 333fred** [reply]
@VSadov has actually been playing a bit with a limited form of dependent types for arrays as part of a recent hackathon: https://github.com/dotnet/roslyn/pull/57286.

**Reply by ufcpp** [reply]
I'd like it to be natively supported. It's too ugly to use ranks of array as integer values instead.

**Comment by huoyaoyuan** [comment]
The overhead of runtime may be huge. The runtime team are avoiding too much generic instantiates.

**Reply by smoothdeveloper** [reply]
What I understand from @OdinAnDieFreude is that it is resolved statically during compilation, the same way generic does when you are not constructing concrete types at runtime.

**Reply by OdinAnDieFreude** [reply]
Exactly, since C# is (except for dynamic) a statically typed language, we definitely want all types to be known at compile time.

And if some value is known at compile time, we could pass it as a generic value parameter to create  a new type.

That would not only allow a lot of options, but also a lot of optimizations, because some checks that were runtime checks could now be performed at compile time.

    public class AddableVector<T, int size> : IAdd<AddableVector<T>> where T : IAdd<T>
    {
      private T<size>[] Data = new T<size>[size]; //array also allows generic value parameters
      public T this[int index] => Data[index]; //depending on the code analysis capabilities, out of bound access to this indexer could be a compile time error
      public AddableVector<T, size> Add(AddableVector<T, size> other)
      {
        var result = new AddableVector<T, size>();
        for (int i = 0; i < size; i++) //C# knows size and can eliminate this loop for low size-values
          result[i] = Data[i].Add(other[i]); //no bounds check needed on Data[i], depending on the analysis also not on the other two.
        return result;
      }
    }
